Don't index themes anymore

Arnei opened the pull request:
Fixes #5264.

Removes all code that relates themes to the index. For services like the admin ui that require detailed query capabilities, a new function has been added to the themes database.

See related issue for motivation.

Of course, this patch cannot remove any actual indices from ElasticSearch/OpenSearch installations. So if you really want themes gone from your index you'll have to nuke and rebuild it, or send a query directly to your index that gets rid of them. Not doing so should not be a problem though.

### How to test this patch

This mostly affects the admin ui and the listprovider, so be sure to check that those work as before.


### Your pull request should…

* [ ] have a concise title
* [ ] [close an accompanying issue](https://docs.opencast.org/develop/developer/#participate/development-process/#automatically-closing-issues-when-a-pr-is-merged) if one exists
* [ ] [be against the correct branch](https://docs.opencast.org/develop/developer/development-process#acceptance-criteria-for-patches-in-different-versions)
* [ ] include migration scripts and documentation, if appropriate
* [ ] pass automated tests
* [ ] have a clean commit history
* [ ] [have proper commit messages (title and body) for all commits](https://medium.com/@steveamaza/e028865e5791)
* [ ] explain why it needs to be merged into the legacy branch, if it is targeting the legacy branch

Arnei committed:
Create db query for themes for admin ui backend

Adds a method to the themes service database that allows to query
the database for themes just like one would query the index (meaning
with limit, offset, sorting and filtering). Also replaces index queries
for themes in the admin ui backend with this new method.

Arnei committed:
Use db instead of index in ThemesListProvider

Instead of querying the index, query the database instead.

Arnei committed:
Remove themes from index

Remove all code that puts themes in the elasticsearch index
or interacts with themes in the index in some other fashion.
The db should be used for theme queries instead.

Arnei committed:
Add test for method "findThemesQuery([...])"

Adds some basic testing to the method for querying themes
from the database. Also fixes a mistake or two that showed
themselves during testing.

KatrinIhler commented on modules/admin-ui/src/test/java/org/opencastproject/adminui/endpoint/TestSeriesEndpoint.java:
```diff
@@ -229,8 +233,29 @@ public TestSeriesEndpoint() throws Exception {
+    EasyMock.expect(userDirectoryService.loadUser((String) EasyMock.anyObject())).andReturn(user).anyTimes();
+    EasyMock.replay(userDirectoryService);
+
+    themesServiceDatabaseImpl = new ThemesServiceDatabaseImpl();
```
Wouldn't it be sufficient for this test to mock this as well?

Arnei replied:
Arguably yes, since in an ideal world the functionality of the themes database has already been tested by its respective module. So we could spare us the trouble of testing it here and save a couple of lines by mocking up the database instead. On the other hand, I see no harm in double testing the themes database here.

KatrinIhler commented on modules/themes/src/test/java/org/opencastproject/themes/ThemesServiceDatabaseTest.java:
```diff
@@ -121,4 +112,96 @@ public void testStoreUpdateAndDelete() throws Exception {
+        Optional.ofNullable("admin")
+    );
+    Assert.assertEquals(1, themes.size());
+  }
```
:heart:

KatrinIhler commented on modules/themes/src/main/java/org/opencastproject/themes/persistence/ThemesServiceDatabaseImpl.java:
```diff
@@ -185,6 +164,99 @@ private List<Theme> getThemes() throws ThemesServiceDatabaseException {
     }
   }
 
+  public List<Theme> findThemesQuery(
```
Nitpick: Just call it findThemes(), bc adding Query implies, for me at least, that you get back a query object, not the results.

KatrinIhler reviewed (changes requested):
Looks pretty good! Love that you extended the tests as well. Just a few minor things, and I stumbled over one issues during testing that already exist in develop: When changing the theme for an existing series, the dropdown shows the ID of the theme after selection, and the series cannot be saved. I know you didn't break this, but could you take a look at that either way?

Arnei committed:
Rename "findThemesQuery" to "findThemes"

To make it clear that this does not actually return a
query object for an index

Arnei commented:
> Issue that already exists in develop

I'd say that is better solved in a different pull request. You're mentioning a dropdown, so might this be an admin-ui issue? I suggest opening a ticket over there https://github.com/opencast/opencast-admin-interface/issues

KatrinIhler commented:
> > Issue that already exists in develop
> 
> I'd say that is better solved in a different pull request. You're mentioning a dropdown, so might this be an admin-ui issue? I suggest opening a ticket over there https://github.com/opencast/opencast-admin-interface/issues

Done, see https://github.com/opencast/opencast-admin-interface/issues/1205

Would be great if you could take a look at that because it kinda gets in the way of properly testing this. :)

Arnei committed:
Merge branch 'develop' into remove-themes-from-index

Arnei committed:
Merge branch 'develop' into remove-themes-from-index

KatrinIhler commented:
Works now! Do you want to rebase this onto 18.x while there's still time?

Arnei commented:
What do you mean by "while there's still time"? The branch cut already happened, right? In any case, I don't think this is important enough to force into 18.
